File: src/main/java/proto/games/matchmaker/domain/matchmaker/PlayerBulk.java
package proto.games.matchmaker.domain.matchmaker;

import lombok.Getter;
import proto.games.matchmaker.domain.player.Player;
import proto.games.matchmaker.domain.player.Rank;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

public class PlayerBulk implements Cloneable {
    private final List<Player> players;

    @Getter
    private Rank medianRank;

    public PlayerBulk(Collection<Player> players) {
        this.players = new ArrayList<>(players);
        Collections.sort(this.players);
        defineMedianRank(this.players);
    }

    public List<Player> takeTopHighLevelPlayers(int length) {
        List<Player> topPlayers = new ArrayList<>(players.subList(0, length));
        remove(topPlayers);
        return topPlayers;
    }

    public List<Player> getPlayers(){
        return new ArrayList<>(players);
    }

    public void remove(List<Player> players) {
        for (Player player : players) {
            remove(player);
        }
    }

    public void remove(Player player) {
        players.remove(player);
    }

    public long count() {
        return players.size();
    }

    @Override
    public PlayerBulk clone() {
        return new PlayerBulk(new ArrayList<>(players));
    }

    private void defineMedianRank(List<Player> players) {
        if (players.size() % 2 == 0) {
            this.medianRank = players.get(players.size() / 2 + 1).getRank();
        } else {
            this.medianRank = players.get(players.size() / 2).getRank();
        }
    }
}

File: src/main/kotlin/proto/games/matchmaker/domain/matchmaker/PlayerBulk.kt
package proto.games.matchmaker.domain.matchmaker

import proto.games.matchmaker.domain.player.Player
import proto.games.matchmaker.domain.player.Rank

class PlayerBulk(players: Collection<Player>) : Cloneable {
    private val _players: MutableList<Player> = players.sorted().toMutableList()
    val medianRank: Rank = defineMedianRank(_players)

    val players: MutableList<Player>
        get() = ArrayList(_players)

    fun takeTopHighLevelPlayers(length: Int): List<Player> {
        val topPlayers: List<Player> = ArrayList(_players.subList(0, length))
        remove(topPlayers)
        return topPlayers
    }

    fun remove(players: List<Player>) {
        for (player in players) {
            remove(player)
        }
    }

    fun remove(player: Player?) {
        _players.remove(player)
    }

    fun count(): Long {
        return _players.size.toLong()
    }

    public override fun clone(): PlayerBulk {
        return PlayerBulk(ArrayList(_players))
    }

    private fun defineMedianRank(players: List<Player>): Rank {
        return if (players.size % 2 == 0) {
            players[players.size / 2 + 1].rank
        } else {
            players[players.size / 2].rank
        }
    }
}